core/sql/optimizer/Inlining.cpp

```diff
@@ -618,25 +619,58 @@ RETDesc *GenericUpdate::createOldAndNewCorrelationNames(BindWA *bindWA, NABoolea
   }
 
   if ((getOperatorType() != REL_UNARY_INSERT) || 
-	getUpdateCKorUniqueIndexKey())
+      getUpdateCKorUniqueIndexKey() ||
+      ((getOperatorType() == REL_UNARY_INSERT) &&((Insert *)this)->isMerge()) ||
+      ((getOperatorType() == REL_UNARY_INSERT) && ((Insert *)this)->isUpsert() && (CmpCommon::getDefault(TRAF_UPSERT_TO_EFF_TREE) == DF_ON )))  
   {
     // DELETE or UPDATE --
     // Now merge the old/target/before valueid's (the Scan child RETDesc)
     // into this RETDesc such that these cols are all named "OLD@.<col>"
-    //
+    //f
     Scan *scan ;
     if (getOperatorType() != REL_UNARY_INSERT)
       scan = getScanNode();
     else 
       scan = getLeftmostScanNode();
+    if ((getOperatorType() == REL_UNARY_INSERT) && ((Insert *)this)->isUpsert() && (CmpCommon::getDefault(TRAF_UPSERT_TO_EFF_TREE) == DF_ON ))
+      {
+	RelSequence *olapChild = getOlapChild();
+	CorrName corrName(getTableDesc()->getCorrNameObj().getQualifiedNameObj(), 
+			  bindWA->wHeap(),
+			  OLDCorr);
+	
+        //	ColumnDescList *colList = (olapChild->getRETDesc())->getColumnList();
+	for (short i = 0; i< olapChild->getRETDesc()->getDegree();i++)
+	  {
+	    // we remembered if the original columns was from the right side of
+	    // this olap node so add those to the RetDesc since those are the 
+	    //ones we want to delete from the dependent indexes.
+	    if ((olapChild->getRETDesc()->getValueId(i)).getItemExpr()->origOpType() == ITM_INSTANTIATE_NULL)
```

Comment:
We are not remembering the origOpType in any other operator. So I think we may be ok with this. Only in the RelSequence operator, we remeber what the original opType was.